src/cleartext.rs

```diff
@@ -128,46 +132,46 @@ impl CleartextSignedMessage {
     ///
     /// On success returns the first signature that verified against this key.
     pub fn verify(&self, key: &impl PublicKeyTrait) -> Result<&StandaloneSignature> {
-        let nt = self.normalized_text();
-        for signature in &self.signatures {
-            if signature.verify(key, nt.as_bytes()).is_ok() {
-                return Ok(signature);
+        if let Some(nt) = self.signed_text() {
+            for signature in &self.signatures {
+                if signature.verify(key, nt.as_bytes()).is_ok() {
+                    return Ok(signature);
+                }
             }
         }
         bail!("No matching signature found")
     }
 
-    /// Verify each signature, potentially agains a different key.
+    /// Verify each signature, potentially against a different key.
     pub fn verify_many<F>(&self, verifier: F) -> Result<()>
     where
         F: Fn(usize, &StandaloneSignature, &[u8]) -> Result<()>,
     {
-        let nt = self.normalized_text();
-        for (i, signature) in self.signatures.iter().enumerate() {
-            verifier(i, signature, nt.as_bytes())?;
+        if let Some(nt) = self.signed_text() {
+            for (i, signature) in self.signatures.iter().enumerate() {
+                verifier(i, signature, nt.as_bytes())?;
+            }
+            Ok(())
+        } else {
+            bail!("Signature can't be verified, message text isn't valid UTF8")
         }
-        Ok(())
     }
 
-    /// Normalizes the text to the format it is used as signature.
-    pub fn normalized_text(&self) -> String {
-        let mut out = String::new();
-        for line in self.text.lines() {
-            // drop dash escapes if they exist
-            if let Some(line) = line.strip_prefix("- ") {
-                out += line;
-            } else {
-                out += line;
-            }
-            out += "\r\n";
-        }
+    /// Normalizes the text to the format that was hashed for the signature.
+    /// The output is normalized to "\r\n" line endings.
+    ///
+    /// Returns `None` if the text isn't valid UTF8
+    pub fn signed_text(&self) -> Option<String> {
+        let unescaped = dash_unescape(&self.csf_encoded_text);
+
+        let normalized: Vec<u8> = Normalized::new(unescaped.bytes(), LineBreak::Crlf).collect();
 
-        out
+        std::str::from_utf8(&normalized).map(str::to_owned).ok()
```

Comment:
you can use `.expect` here, it is guranteed that this is valid utf8, because `csf_encoded_text` already is

Reply:
great, done